### DISCUSSION TITLE
How to masking money in filament 3 ?

### DISCUSSION THREAD

**Discussion by jigujigu** [filamentphp/filament#9069] [Help] [answered]
Hi All. I got error when masking textinput for money in filament 3

```
use Filament\Forms\Components\TextInput\Mask;
....
  TextInput::make('price')
      ->mask(
          fn (Mask $mask) => $mask
              ->numeric()
              ->decimalPlaces(2)
              ->thousandsSeparator(',')
              ->decimalSeparator('.')
      ),
```
**Target class [Filament\Forms\Components\TextInput\Mask] does not exist.**

Thank you.

**Comment by sinnbeck** [comment, marked as answer]
Sadly there is no Mask class anymore

https://filamentphp.com/docs/3.x/forms/fields/text-input#input-masking

But you could try copying it into your own project and see if it works still
https://github.com/filamentphp/forms/blob/2.x/src/Components/TextInput/Mask.php

Or you can use the new RawJs helper and just directly call the alpine money mask
https://alpinejs.dev/plugins/mask#money-inputs
```php
TextInput::make('price')
      ->mask(RawJs::make(<<<'JS'
$money($input, '.', ',', 2)
JS
                    ))

**Reply by jigujigu** [reply]
Thanks

**Reply by luiseduardobraschi** [reply]
That just doesn't seem to work. The rendered code from my Filament field definition looks fine:

![image](https://github.com/filamentphp/filament/assets/700448/ca0be6d6-3c58-447b-ae3d-db045a24241a)

As per Alpine's doc example, it should mask "1234" to "12,34", but all I got is this instead:

![image](https://github.com/filamentphp/filament/assets/700448/30712a87-7e9f-4a24-ad77-5b265430ac3e)

**Reply by hfoletto** [reply]
> That just doesn't seem to work. The rendered code from my Filament field definition looks fine:
> 
> ![image](https://private-user-images.githubusercontent.com/700448/329711092-ca0be6d6-3c58-447b-ae3d-db045a24241a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUxNjk0NzMsIm5iZiI6MTc0NTE2OTE3MywicGF0aCI6Ii83MDA0NDgvMzI5NzExMDkyLWNhMGJlNmQ2LTNjNTgtNDQ3Yi1hZTNkLWRiMDQ1YTI0MjQxYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyMFQxNzEyNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wM2Q0NDY4YWE4ODA1Y2EyNWUyNDgxZDNhYjczOWVjODI1NzEwZjY2NjhlYzFiNTQ0YTA5ZjY0ZjdmYTllNDllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2D9p7UPm9EbKfSrTHwKE97L1PN90x3KmotEfrJjdADc)
> 
> As per Alpine's doc example, it should mask "1234" to "12,34", but all I got is this instead:
> 
> ![image](https://private-user-images.githubusercontent.com/700448/329711843-30712a87-7e9f-4a24-ad77-5b265430ac3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUxNjk0NzMsIm5iZiI6MTc0NTE2OTE3MywicGF0aCI6Ii83MDA0NDgvMzI5NzExODQzLTMwNzEyYTg3LTdlOWYtNGEyNC1hZDc3LTViMjY1NDMwYWMzZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyMFQxNzEyNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMTFkYzZmYzExMDQyNDA4NjAyYzllZGRjZjgwYmMxOThkYTFkNjA3OWViODI2YTU4NDJjMjVhZTcwM2YwNWY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.99zq5lDpff2iYtR6kWNpwvGAF4q9aGn1H8KfRa6fzls)

@luiseduardobraschi I'm facing the same problem. Did you ever figure this out?

**Reply by luiseduardobraschi** [reply]
@hfoletto No. Not just using "::mask()". I ended up using my very own "hacky" code.

**Comment by wickham-c** [comment]
At the end of the [Filament documentation](https://filamentphp.com/docs/3.x/forms/fields/text-input#input-masking) section @sinnbeck referred to, there's a very useful code snippet specific to money-handling which worked nicely for me...
```
use Filament\Forms\Components\TextInput;
use Filament\Support\RawJs;
 
TextInput::make('amount')
    ->mask(RawJs::make('$money($input)'))
    ->stripCharacters(',')
    ->numeric()
```

**Comment by softwaretechniker** [comment]
I've comma as decimal separator as it is the local formatting and store it as an integer in the smallest unit (cents) to avoid round-off errors in any kind of calculations (what is actually not that rare as some may think).

The solutions mentioned above didn't work for this case and I ended up doing the following:
```
$decimalStateFormating = fn ($state) => ($state == '') ? '' : str_replace('.', ',', $state / 100);
$moneyMask = RawJs::make("\$money(\$input, ',', '.', 2)");
//…
TextInput::make('replacement_value')
    ->label(static::__('attribute.replacement_value'))
    ->formatStateUsing($decimalStateFormating)
    ->mask($moneyMask)
    ->prefix('€'),
```

Plus setting up some cast and add it the model for writing:
```
$value = (string) $value;
$decimalPos = strrpos($value, $this->decimalSeparator);
$targetLength = $decimalPos + 1 + $this->decimals;
return (int) str_replace(
    [$this->thousandSeparator, $this->decimalSeparator],
    '',
    str_pad(substr($value, 0, $targetLength), $targetLength, '0')
);
```
A pure client (JS) (or at least UI / Filament) solution would be cleaner than the cast on Eloquent level, but I'm not sure how to intercept Filament's form submit...

In the table column definition, the following does all the magic:
```
TextColumn::make('replacement_value')
    ->label(static::__('attribute.replacement_value'))
    ->sortable()
    ->money('EUR', divideBy: 100),
```

**Reply by joaovitoralvares** [reply]
For formating on the client side, you should use formatStateUsing and dehydrateStateUsing methods.
So for formating the value coming from the back-end, you should use formatStateUsing. For formating client-side value before sending to backend, you should use dehydrateStateUsing.

You can get more information about this on th following section:
https://filamentphp.com/docs/3.x/forms/advanced#field-lifecycle

**Reply by begueradj** [reply]
That is not useful for input validation/formatting  @joaovitoralvares
